src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java

```diff
@@ -1090,7 +1092,10 @@ private void createDependencyReducedPom( Set<String> artifactsToRemove )
             origDeps = transitiveDeps;
         }
 
-        Model model = project.getOriginalModel();
+        final Model model = project.getFile() == null
```

Comment:
looks like an issue any plugin consuming the pom will get so it should likely be fixed in flatten plugin using `setOriginalModel` to be a global fix and not a local one.

Reply:
hi @rmannibucau , the in Maven, what is 'originalModel` defined to be, a model of the original pom or a model of the current pom file? If defined as the latter one, the global solution of updating originalModel could be done in `MavenProject::setPomFile` or `MavenProject::setFile`.

Reply:
My guess is that it was the really original file but since maven is a chain (per module) it is in practise the current pom file - otherwise it can't really be used by any plugins and some do.
So guess the choice is:

1. keep original untouched but means no plugin should use it except help/debug ones (so fix is to always read getFile())
2. ensure flatten plugin updates original model to let downstream plugins consumes the post processing model

Personally I think the option 2 is what makes the most sense for the ecosystem (inter-plugin operability).

Reply:
ok, this pr is the first choice, which prefer a practice single source to get current raw model.

Reply:
@gzm55 this pr does not impl 1 since it should never use original model in case 1 but always file to enable to chain plugins.

Reply:
it is ok to not fall back to original model when `getFile()` is null, throwing  an exception instead.

Reply:
Hmm, I just checked flatten code and it only works if `updatePomFile` is set to true no? So long story short, if you want it to work reliably you must ensure originalModel is set before shade plugin is executed. From a design point of view it is the responsability of flatten plugin to do that since it is the one hacking the user pom but an intermediary plugin setting it from a file when flatten plugin dumped its value would work too but I don't think there is any bug in shade there.

Reply:
it seems better to have a convention or rule that the original model should be equal to the pom file. the shade plugin also breaking this convention, it reverts the parent.relativePath after write the model to the file.

Reply:
shade plugins reset it cause it just needs to change it while dumping the pom (that said good point it breaks thread safety in concurrent builds).
agree shade plugin breaks this rules and happy to get it fixed, it is just not that important cause shade plugin is generally in the end of the chain (so dowstream consumers list is empty) and it also handles itself the pom rewriting but you are right.
long story short: I don't think we should re-read from the filesystem any pom model anytime in maven build otherwise the design itself is broken so I'm in favor of trying to not use the reader more than once per file.

hope it makes sense.